homework/H3_Custom_Types.jl

```diff
@@ -692,9 +695,71 @@ Markdown.MD(
     ),
 )
 
+# ╔═╡ cfcd4e18-efe6-49c5-82b0-e1a1db394935
+md"Please note that optional tasks offer *extra* points. These *extra* points can help compensate for any missed points across all homework submissions, contributing towards the maximum achievable score of 20 points for all combined homework assignments. However, the total score from homework submissions cannot exceed 20 points."
+
+# ╔═╡ fde75269-bfe7-4430-b312-7274c53c7369
+md"### Exercise 3 – Abstract types and Subtypes"
+
+# ╔═╡ 403864a6-faf1-4e5b-b352-76a94cad8248
+md"""
+In this task, you will define an abstract type for a generic shape and create specific subtypes for different shapes such as `Circle`, `Rectangle`, and `Square`. You will also implement a method to calculate the area of each shape."""
+
+# ╔═╡ 4f1a47f9-cd36-4166-9489-7f30f517b513
+task(
+    md"""
+### Step 1: Define the abstract type
+Create an abstract type called `Shape`. 
+### Step 2: Define the subtypes
+Create the following subtypes:
+* `Circle` with a field for the radius
+* `Rectangle` with fields for width and height
+* `Square` with a field for the length of its sides
+### Step 3: Implement the `area` method
+Write a method `area(shape::Shape)` that calculates the area of the shape based on its type. The method should work for all defined subtypes.
+
+The area formulas are as follows:
+ - For `Circle`: $\text{Area} = \pi \times r^2$
+ - For `Rectangle`: $\text{Area} = \text{width} \times \text{height}$
+ - For `Square`: $\text{Area} = \text{side}^2$
+""", 2
+)
+
+# ╔═╡ d708f8f2-4af4-4337-a2e4-f67c0f335de9
+Markdown.MD(
+    Markdown.Admonition(
+        "warning",
+        "Note",
+        [
+            md"This exercise does not provide automated feedback in order to get you to practice programming in Julia on your own!",
+        ],
+    ),
+)
+
+# ╔═╡ 60619a5a-e8f5-40d3-b1ef-6ded4ee9605f
+function area(shape::Shape) # Don't change this line
+	# Write your code here
+
+	return missing
+end
+
+# ╔═╡ c05ef2bc-8bc6-43d8-84aa-d780283af939
+md"The following code segments should help you test your implementation. It is recommended to extend this for the other shapes."
+
+# ╔═╡ 4a3f2458-28f4-4b7c-b68d-eeee67cb2b78
+c = Circle(5)
+
+# ╔═╡ fbb40861-3905-4b83-b762-4a0ba4ccb20c
+area_c = area(c)
+
+# ╔═╡ 01efb844-d95f-4b9f-bc61-153060f561c0
+hint(
+    md"Use `isa` to differentiate the different Subtypes in `area`",
```

Comment:
I thought we wanted to remove this type of feedback-less bonus point exercise because it is takes a lot of effort to grade?

Also it isn't ML/DL themed.

Reply:
Makes sense, yes. After remove this task, HW3 will only have 8 points. Compared to other assignments, it feels liks HW3 has fewer tasks, which isn't necessarily a bad thing. We can, instead, increase the points for the existing tasks.

Reply:
I would say we remove this and either (a) just increase the points or (b) make ADAM mandatory and add tests for it.

Reply:
I think option (a) makes more sense. I propse we increase points for the structs in general, in task 1 and task 2.